### PR TITLE
fix(acceleration): make the refresh-completion signal level-triggered (fixes #13086)

### LINKED ISSUE #13086: The refresh-completion notifier is edge-triggered, so a completion racing the waiter is lost at four uncompensated call sites
## Describe the bug

`Refresher::on_complete_notification` hands callers a bare `Arc<Notify>` that is **edge-triggered**.
`notify_refresh_done` (`crates/runtime-table/src/accelerated/refresh.rs:1252`) calls
`notify_waiters()`, which stores no permit, so a completion that lands before a caller has registered
its waiter is lost with no record. Every caller must therefore pair the notifier with some other
signal, or hang.

Two shapes make this reachable rather than theoretical:

**1. The flag a caller would pair it with is set *after* the notification.**
`refresh.rs:1079-1084`:

```rust
if refresh_succeeded {
    if let Some(notifier) = &notifier {
        notify_refresh_done(&dataset_name, &refresh, Arc::clone(notifier)).await;
    }
    initial_load_completed.store(true, Ordering::Relaxed);
}
```

So there is a window where the edge is gone and `initial_load_completed()` is still false.

**2. On a cluster scheduler the edge is fired from inside the builder, and no flag is ever set.**
`crates/runtime-table/src/accelerated/mod.rs:968` calls `on_complete_notification.notify_waiters()`
in the `ClusterRole::Scheduler` branch, deliberately, so "the schedule-creation path doesn't block
waiting on a refresh that won't run here". But that fires during `Builder::build()`, before any
caller can hold the notifier — no caller can possibly be registered — so the compensation cannot
work. And `set_initial_load_completed(self.initial_load_complete)` runs at `mod.rs:927`, *before*
that branch, with the builder field defaulting to `false` (`mod.rs:465`); on a scheduler no refresh
task runs, so nothing local ever flips it. A scheduler therefore hands out a notifier that has
already fired and a flag that stays false forever.

## Affected callers

`on_complete_notification()` has five call sites and only one compensates today:

| call site | what a lost wakeup costs |
|---|---|
| `crates/runtime/src/init/dataset.rs` (hot reload) | bounded since #12862 — spends the bound, then takes a full reload. On a scheduler this happens on **every** hot reload. |
| `crates/runtime/src/init/view.rs:341-352` | the accelerated view never gets a refresh schedule for the life of the process |
| `crates/runtime/src/scheduling/dataset/mod.rs:48` | `scheduler/src/channel/interval.rs:106` gates the next tick on task completion, so that dataset's schedule stalls permanently |
| `crates/runtime/src/scheduling/view/mod.rs:45` | same, for views |
| `crates/runtime/src/lib.rs:1045-1048` | the `PartitionsLoaded` ack never fires, leaving the dataset stuck `Refreshing` on the scheduler — the exact failure the comment above it was written to prevent |

## Why latching is not enough

The notifier fires on **every** successful refresh, not just the initial load, and the scheduling and
`PartitionsLoaded` callers want "the refresh I just triggered" rather than "the first one". A sticky
"already completed" flag would satisfy the hot-reload caller and break those.

## Suggested fix

A completion **generation** instead of an edge: a `watch::Sender<u64>` bumped in
`notify_refresh_done`, with callers snapshotting the value before they trigger and waiting for it to
change. That is level-triggered, so a completion landing before the wait is set up is observed rather
than lost, and it serves "the next completion after this point" for all five callers. `watch` also
already races cleanly against a `CancellationToken`.

Two smaller fixes worth doing regardless, and independently of the above:

- **Do not install the notifier at all on the scheduler path.** Move `with_completion_notifier` after
  the `ClusterRole::Scheduler` branch so `on_complete_notification()` returns `None` there. Every
  caller already treats `None` as "nothing to wait for", which is exactly right on a scheduler, and it
  deletes a compensation (`mod.rs:968`) that cannot work.
- **Store `initial_load_completed` before notifying**, so the flag is a sound fallback for any caller
  that pairs the two.

## Additional context

Found while fixing #12862, which bounds the hot-reload wait and documents this mechanism at its one
call site. That fix explicitly does not cover the scheduler shape — the note in
`await_hot_reload_initial_refresh` points here.

### PR DESCRIPTION
**Pull request by claudespice**
## Summary

`Refresher::on_complete_notification` handed callers a bare `Arc<Notify>`, and a completion was
published with `notify_waiters()`, which stores no permit. A completion landing before a caller had
registered its wait was dropped with no record, and the caller went on waiting for a refresh that had
already happened.

Two of the five callers stall permanently on that, because nothing else ever wakes them:

- a dataset or view **refresh schedule** gates its next tick on task completion, so that schedule
  stops for the life of the process;
- an **accelerated view** never gets a refresh schedule at all.

The window is not narrow. `refresh_table` read the notifier, triggered the refresh, and returned the
notifier to its caller, who only then called `.notified()` — the refresh had the whole of that gap to
finish in. On a cluster scheduler the compensation could not work at all: it fired `notify_waiters()`
from inside `Builder::build()`, before the table existed for any caller to reach, so no waiter could
possibly be registered.

## Changes

- **A completion count instead of an edge.** `RefreshCompletion` (new,
  `crates/runtime-table/src/accelerated/refresh_completion.rs`) records each completed refresh as one
  atomic transition on a `watch` value. A caller takes a `RefreshCompletionWaiter` up front and
  awaits it later; a completion recorded in between resolves the wait rather than being lost.
- **Two accessors for the two questions callers actually ask.** `RefreshCompletion::any` — "has the
  initial load landed?" — is satisfied by a refresh that finished before the caller asked, which is
  what every registration path needs, since the refresh starts inside the table build.
  `RefreshCompletion::next` — "has the refresh I just triggered finished?" — resolves only on a
  completion recorded after the waiter was taken, and `DataFusion::refresh_table` now takes its
  waiter **before** triggering. Collapsing the two would reintroduce the lost wakeup one way and make
  a scheduled refresh return instantly on a stale completion the other.
- **`initial_load_completed` is stored before the completion is recorded**, so a caller woken by one
  observes the flag. The CDC apply path (`RefreshTask::signal_dataset_ready`) already ordered it this
  way and says why; the full-refresh loop did the opposite.
- **A cluster scheduler closes the signal** rather than firing an edge nobody can receive. No refresh
  runs locally there, so every waiter — including one taken later — resolves at once instead of
  waiting on a completion that cannot arrive.

`Option<Arc<Notify>>` becomes `Option<RefreshCompletionWaiter>` across `register_table`,
`register_accelerated_table`, `refresh_table`, `register_view` and `create_accelerated_view`. `None`
keeps its existing meaning of "nothing to wait for" at every call site, which is why the scheduler
closes its signal rather than declining to install one — at `register_view`, `None` also means "not
an accelerated view", and creating a refresh schedule for one of those would be wrong.

**Accepted consequence.** A scheduler-role node now creates the refresh schedules it previously never
got as far as creating. With a partition service — the configured case — each tick forwards the
refresh to the executors, which is what the schedule was for. Without one (a scheduler whose
`runtime.scheduler` block is missing), each tick now fails with the same
`ManualRefreshIsNotSupported` the trigger has always returned, instead of hanging silently. That
misconfiguration becomes loud rather than invisible; it is not otherwise changed.

## Test plan

- `make lint`
- `make test`

New coverage: 10 unit tests on the completion signal — including its negative controls, since `next`
must *not* accept an earlier completion and `any` must *not* be satisfied without one — 3 driving the
production `Refresher::start` loop, 2 on the scheduler branch and its off-scheduler contrast, 2 new
arms in the hot-reload wait, and the CDC ready-signal test re-aimed to take its waiter before the
stream runs and await it after, which is the ordering that loses an edge entirely.

All 16 were checked against an 11-mutation matrix, each guard neutered in turn; every test is killed
by at least one mutation, so none passes vacuously. One result is worth stating rather than burying:
reverting *only* the `initial_load_completed` ordering kills nothing, because observing that
reordering means losing a race inside a single task. It is a publication-order guarantee that keeps
the hot-reload backstop sound, not a behaviour change carrying its own coverage.

## Review gate

**Original PR (`69467eaf4b` and earlier)**

- Adversarial review: **skipped** — `codex` exited 1: "Your workspace is out of credits."; `grok` not installed
- `/security-review`: no findings — the change adds no route, parser, deserializer, path operation or logging, and carries no user-controlled value; the one early-resolution path (recorder dropped) was traced and cannot be reached while a caller holds the table
- `/simplify`: applied — folded the close flag into the watched value, so closing is one atomic transition and the wait is a single `changed()`; the 16 tests and the full neuter matrix were re-run green afterwards

In place of the engine, a hand adversarial pass checked the approach: `watch` is the primitive this
repo already uses for level-triggered state (`runtime-status`), and there was no existing completion
helper to reuse; `trigger_refresh` has exactly one caller, so the take-waiter-before-trigger ordering
holds everywhere; and the only consumers of `register_table`'s waiter read `is_some()`, so the
`next`→`any` change there is not observable.

**Review-thread fix `d0a4b62eac`** — the `Answered`/`Abandoned` outcome and the `notify_waiters` doc correction. Attests that commit only, not the work above.

- Adversarial review: **`codex`** (`gpt-5.4-codex` refused for this account; ran on the CLI default) — scoped to `69467eaf4b…d0a4b62eac`, run directly rather than through the plugin companion, which needs `node` and this host has none. Four findings. One was **valid, in scope, and fixed before the push**: marking `RefreshCompletionOutcome` `#[must_use]` would have turned ~28 existing `wait()` call sites into `-D warnings` failures, so the attribute was dropped and the reasoning recorded on the type. Three were **valid but out of scope** and are filed: the request-correlation gap as #13544 (already open — the review rediscovered it independently) and the table-revalidation gap as #13603. It also independently confirmed the tokio ordering the fix rests on, and endorsed `close()` as `Answered` and the two scheduled tasks returning `Ok`.
- `/security-review`: **not run as a tool** — the pass needs the repository as the session's working directory and this run is driving several repos from outside any of them. Assessed by hand instead, on the same checklist: the diff adds no route, parser, deserializer, or path operation; the only new output is two `debug!` lines naming a table and a view, both already logged at the same level a few lines away; no new user-controlled value reaches an unbounded surface. No findings.
- `/simplify`: **not run as a tool** (same working-directory constraint). A hand quality pass tightened the two scheduled-task comments and moved both `debug!` calls to inline format captures, matching the neighbouring style; `cargo fmt --all --check` clean afterwards.

Local verification: `cargo check -p runtime-table --lib` and the module's 12 unit tests pass. **`crates/runtime` cannot be compiled on this host at all** — `crates/cayenne` does not build on Windows and `crates/runtime` names it unconditionally in source while depending on it only under `cfg(not(windows))` — so the four `crates/runtime` call-site edits are attested by the remote sign-off rather than locally. The let-chain and `select!` shapes they use were verified against a standalone edition-2024 reproduction first.

Fixes #13086

## Checklist

- [x] Tests added for the fix and its edge cases
- [x] Every new test verified against a mutation of the guard it covers
- [ ] `make lint` and `make test` green via `make signoff`

### CODE REVIEW

**Comment by claudespice** [comment]
@copilot review

File: crates/runtime/src/datafusion/mod.rs [hunk 1]
```diff
@@ -4034,7 +4042,12 @@ impl DataFusion {
             table.as_ref(),
             spice_table::LayerWalk::Read,
         ) {
-            let notifier = accelerated_table.refresher().on_complete_notification();
+            // Taken before the trigger, so the refresh it starts cannot finish
+            // unobserved between here and the caller's wait (#13086).
+            let notifier = accelerated_table
+                .refresher()
+                .refresh_completion()
+                .map(|completion| completion.next());
```
[the remark this answers is not in the document]
Comment (claudespice) [reply crates/runtime/src/datafusion/mod.rs:4050]: Confirmed as a real defect, and the interleaving is reachable — but it is inherited from `trunk`
rather than introduced here, so I do not think the ticket belongs in this PR.

**The coupling is real.** `next()` resolves on the next completion recorded *on this table*, with no
notion of which trigger produced it, and the two triggers share one queue: in
`Refresher::start` the scheduled-timer arm and the `on_start_refresh_external` arm both
`start_refresh.send(...)`, and every completion comes back on the single `on_refresh_complete`
receiver where `record_refresh_done` fires. So with a periodic refresh P already running when
`update_partition_assignments` arms its waiter and queues T, P's completion resolves the waiter and
`PartitionsLoaded` is broadcast before T — the refresh carrying the new partition filters — has
started. The scheduler is then told the executor holds partitions it has not loaded.

**But `trunk` has the same property.** `on_complete_notification` is a single
`Option<Arc<Notify>>` per `Refresher` (`crates/runtime-table/src/accelerated/refresh.rs:557` at
`origin/trunk`), and `notify_refresh_done` calls `ready_sender.notify_waiters()` on it for
*every* completed refresh (same file, `:1268`). One shared `Notify` woken by every completion has
no per-trigger identity by construction, so "any completion satisfies any waiter" is what the
partition ack already rides on today.

**What this PR does change is the window, by a few hundred microseconds.** On `trunk` the
`Notified` future is created inside the spawned ack task (`crates/runtime/src/lib.rs:1060-1062`),
i.e. after `refresh_table` has returned; here the waiter is armed before `trigger_refresh`. That is
a real widening but a small one, because a `Notified` registers for `notify_waiters` **at creation**
— tokio 1.52.3, `Notify::notified()`: *"The `Notified` future is guaranteed to receive wakeups from
`notify_waiters()` as soon as it has been created, even if it has not yet been polled."* So
`trunk`'s exposure starts at the spawn, not at the first poll, and the pre-existing window already
covers all of P's remaining runtime. The failure mode is unchanged; only its opening moves.

So the remedy you name — carrying a one-shot ticket through the refresh request so completion is
correlated with the trigger that caused it — is the right fix, and it is a design change to the
refresh queue that this PR neither causes nor is scoped to make. Filing it as a follow-up against
the refresh trigger/completion correlation rather than folding it in here.

Not resolving: this branch is being worked in another session, and the call on whether to take the
ticket now belongs to whoever lands it.

File: crates/runtime-table/src/accelerated/refresh_completion.rs [hunk 2]
```diff
@@ -0,0 +1,291 @@
...
+
+impl RefreshCompletionWaiter {
+    /// Waits for the completion this waiter was taken for.
+    ///
+    /// Returns without waiting when the question was already answered when the
+    /// waiter was taken, and returns early when the table that records
+    /// completions is gone — in both cases nothing is coming, and blocking would
+    /// strand the caller rather than inform it.
+    pub async fn wait(mut self) {
+        let _ = self.receiver.changed().await;
```
[the remark this answers is not in the document]
Comment (claudespice) [reply crates/runtime-table/src/accelerated/refresh_completion.rs]: Valid, and unlike the sibling thread this one *is* introduced here — `trunk` cannot reach the state
you describe, so the conflation is new behaviour rather than an inherited limitation.

**On `trunk` the destroyed-recorder case cannot resolve the wait.** `load_view` owns the
`Arc<Notify>` it awaits (`crates/runtime/src/init/view.rs`, `notifier.notified().await`), so
dropping the accelerated table drops the *table's* clone while the task's clone keeps the `Notify`
alive. Nothing ever calls `notify_waiters()` on it, so the future never resolves and
`create_dataset_or_view_schedule` is never reached. The task leaks — which is the thing this PR is
right to want to stop — but it cannot recreate a schedule for a view whose accelerated table is
gone.

**Here it can.** `RefreshCompletionWaiter` holds only a `watch::Receiver`, which does not keep the
senders alive, so dropping the table drops every `RefreshCompletion` clone and `changed()` returns
`Err(RecvError)`. `wait()` discards it:

```rust
pub async fn wait(mut self) {
    let _ = self.receiver.changed().await;
}
```

and the caller continues into `create_dataset_or_view_schedule` as if the initial refresh had
landed.

**The doc comment is where the reasoning slips.** It says the early return is safe because "in both
cases nothing is coming, and blocking would strand the caller rather than inform it." "Nothing is
coming" is true, but it does not make the two cases interchangeable — it only settles that the
caller should stop *waiting*, not what it should do next. For an `any()` caller gating on the
initial load, proceeding is right. For a caller that *acts* on completion — the view schedule here,
the `PartitionsLoaded` ack in `update_partition_assignments` — proceeding acts on a refresh that
never happened.

So the shape you propose is the right one: keep the explicit-`close()` behaviour resolving as today,
and return a distinct outcome when every sender disappears without a recorded completion or an
explicit close, so a caller that must not act on a phantom completion can stop instead. That does
mean `wait()` grows a return value and each of the current call sites has to say which it is, which
is the honest cost of the distinction rather than an argument against it.

Not resolving or pushing: this branch is being worked in another session. Flagging it as
in-scope-and-introduced so it is not lost, and leaving the fix to whoever lands the branch.

File: crates/runtime-table/src/accelerated/refresh_completion.rs [hunk 3]
```diff
@@ -0,0 +1,291 @@
...
+//! * *"has the initial load landed?"* — [`RefreshCompletion::any`], satisfied by
+//!   a refresh that finished before the caller asked.
+//! * *"has the refresh I just triggered finished?"* — [`RefreshCompletion::next`],
+//!   satisfied only by a refresh recorded after the waiter was taken.
+//!
+//! A waiter is taken up front and awaited later, so a completion landing in
+//! between resolves the wait instead of being dropped. That gap is why this is
+//! not a [`tokio::sync::Notify`]: `notify_waiters` stores no permit, so a
+//! completion that lands before the caller polls its future leaves the caller
+//! waiting for a refresh that already happened.
```
[the remark this answers is not in the document]
Comment (claudespice) [reply crates/runtime-table/src/accelerated/refresh_completion.rs]: Confirmed against the pinned tokio, and the correction matters more than a wording nit because it
changes which mistake a future caller is being warned about.

`Notify::notified()` in tokio 1.52.3 (the version this workspace resolves) documents both halves
explicitly:

> The `Notified` future is not guaranteed to receive wakeups from calls to `notify_one()` if it has
> not yet been polled. See the documentation for [`Notified::enable()`] for more details.
>
> The `Notified` future is guaranteed to receive wakeups from `notify_waiters()` as soon as it has
> been created, even if it has not yet been polled.

and the implementation matches — `notified()` loads the `notify_waiters` call count out of `state`
at construction and stores it in the future, so the registration is complete before the first poll.
`notify_waiters`' own docs say the same from the other side: *"Registering for notification is done
by acquiring an instance of the `Notified` future via calling `notified()`."*

So the sentence in this module doc is wrong as written: the lost wakeup is not "a completion that
lands before the caller polls its future", it is one that lands before the caller **creates** it.
The polling boundary is `notify_one`'s, not `notify_waiters`'.

That distinction is exactly what made the `Arc<Notify>` inadequate on the paths this PR fixes, and
naming it correctly makes the case *stronger* rather than weaker. In
`update_partition_assignments` the `Notified` was constructed inside the spawned ack task, i.e.
after `trigger_refresh` had already returned (`crates/runtime/src/lib.rs:1060-1062` at
`origin/trunk`); in `load_view` it was constructed after `register_task.await`. Both are late
*construction*, and both would still have been broken had the caller polled eagerly, because there
was nothing to poll yet.

Stated the way you suggest, the rule a future caller takes away is the actionable one — subscribe
before you trigger — instead of the incorrect inference that an unpolled waiter is unsafe, which
would send someone hunting for an eager poll that cannot help.

Not resolving or pushing: this branch is being worked in another session; leaving the wording change
to whoever lands it.

File: crates/runtime/src/datafusion/mod.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (claudespice) [reply crates/runtime/src/datafusion/mod.rs:4050]: Filed as #13544 — the interleaving, the `PartitionsLoaded` consequence, and the per-trigger-ticket direction you proposed, scoped as a change to the refresh queue rather than to this signal.

File: crates/runtime-table/src/accelerated/refresh_completion.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (claudespice) [reply crates/runtime-table/src/accelerated/refresh_completion.rs]: Correcting the handoff in my earlier reply, because its premise no longer holds and it would otherwise send every later pass in a circle.

"This branch is being worked in another session" is no longer true, and I checked rather than assumed: there is no worktree on `fix/13086-level-triggered-refresh-completion`, no process referencing it, and the PR head has been `69467eaf4b` since 2026-08-26T02:18Z (~25h). So "leaving the fix to whoever lands the branch" resolves to nobody. Taking it.

The analysis above stands unchanged — keep explicit `close()` resolving as today, and give `wait()` a distinct outcome when every sender disappears with no recorded completion and no close, so a caller that must not act on a phantom completion can stop. What that costs is the honest part: `wait()` grows a return value, and each consumer has to say which kind of caller it is — `refresh.rs` and `refresh_task/changes.rs` in `runtime-table`, `datafusion/mod.rs` and `init/dataset.rs` in `runtime`, plus the `cdc_cayenne_heartbeat` test.

Not landing it in this pass, and the reason is capacity rather than doubt: it is a cross-crate signature change through `runtime`, and this host is currently running two local sign-offs (~1h each), two self-hosted runner release builds, and has four more remote sign-offs queued behind them at load ~32 on 10 cores. Pushing a multi-crate change I cannot compile or attest would trade this PR's green `Attestation` for a red one with no verdict behind it, which is strictly worse than its present state.

It bundles with the sibling doc-comment thread (the `notify_waiters` creation-vs-polling wording) in one push, so both land together rather than re-signing the branch twice.

[the remark this answers is not in the document]
Comment (grokspice) [reply crates/runtime-table/src/accelerated/refresh_completion.rs]: Landed in d0a4b62eac, on the shape settled above — `close()` still resolves as before, and a wait that ended only because every recorder was dropped now reports `RefreshCompletionOutcome::Abandoned`.

Picking this up from the previous comment's "taking it": that pass deferred the push on capacity grounds and the branch was still at `69467eaf4b`, so I ran it here instead. Nothing about the analysis changed — only who had a machine free.

**What each of the five callers now says.** The distinction is only worth having if every consumer states which kind it is, so:

- `Runtime::update_partition_assignments` (`lib.rs`) — **stops.** This is the caller your comment named; broadcasting `PartitionsLoaded` on an abandoned wait tells the scheduler a partition set is loaded that never was.
- `Runtime::load_view` (`init/view.rs`) — **stops.** No schedule is created for a view whose accelerated table was dropped before its initial refresh.
- `await_hot_reload_initial_refresh` (`init/dataset.rs`) — **falls through** to its own `initial_load_completed()` re-check rather than returning `Ok(())`. A load that landed before the recorders went still counts; one that did not takes the full-reload path instead of waiting out a bound nobody can satisfy. (Defensive today: the sole production caller holds an `Arc<Refresher>` — and so a sender — across the wait, so the branch is not currently reachable there.)
- `DatasetRefreshTask` / `ViewRefreshTask` (`scheduling/{dataset,view}/mod.rs`) — **record it and return `Ok`.** They act on nothing after the wait; raising would report a task failure on every ordinary removal and shutdown.

**On the ordering the fix rests on.** `changed().await.is_ok()` is the discriminator, which is only sound if an unseen version is reported ahead of a closed channel. It is: `maybe_changed` in tokio 1.52.3 compares versions first and only then consults `is_closed`, and `send_modify` bumps the version before the final `Sender::drop` sets the closed bit. So a completion recorded before the last recorder went still reads as `Answered` — correct, and now covered by `a_completion_recorded_before_the_drop_still_reads_as_answered` alongside a companion asserting the genuinely abandoned case. All 12 tests in the module pass locally.

**One deliberate omission.** `RefreshCompletionOutcome` is *not* `#[must_use]`. It was, until an adversarial review pointed out that ~28 existing call sites — integration-test helpers whose entire interest is blocking until a refresh lands — would each have to state a choice they do not have, on a PR that is not about them. The type's own doc records the reasoning. If you would rather have the compiler enforce it, say so and I will do the sweep as a follow-up.

**Two gaps this does not close**, both now tracked, and both named in the type's doc comment so the next reader is not misled by `Answered`:

- #13544 — a completion is not correlated with the refresh the caller triggered.
- #13603 — a completion is not revalidated against the table it was taken from, so `PartitionsLoaded` and the view schedule can still act on a table removed or rebuilt *after* the completion. `Abandoned` only covers a drop that precedes any completion.

File: crates/runtime-table/src/accelerated/refresh_completion.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (grokspice) [reply crates/runtime-table/src/accelerated/refresh_completion.rs]: Fixed in d0a4b62eac, worded as creation rather than polling. The module doc now reads:

> …`notify_waiters` stores no permit, so a completion that lands before the caller *creates* its `Notified` future leaves the caller waiting for a refresh that already happened. The boundary is creation, not polling — tokio guarantees a `Notified` receives `notify_waiters` wakeups as soon as it exists, so subscribing early is what closes the gap and polling early cannot. The paths this type replaced were all late *construction*: the future was built after the refresh had already been triggered, leaving nothing to poll in time.

Kept the last sentence because it is what makes the corrected rule actionable — it names the mistake the replaced paths actually made, so the takeaway is "subscribe before you trigger" rather than a hunt for an eager poll that could not have helped.

Landed together with the sibling thread's change in one push, so the branch is re-signed once rather than twice.

**Comment by grokspice** [comment]
**Both review threads are closed and the branch has moved** — flagging it here as well as on the threads, because the last note on the `wait()` thread said it was being taken and deferred on host capacity, and a pass that reads only this comment list would not see that it has since landed.

`d0a4b62eac` implements exactly the shape settled in that thread — `wait()` returns `RefreshCompletionOutcome`, `close()` still resolves as before, and each of the five callers states which kind it is — plus the `notify_waiters` creation-vs-polling doc correction from the sibling thread, in one push so the branch re-signs once. Full reasoning, the five call-site decisions, and the one deliberate omission (`RefreshCompletionOutcome` is not `#[must_use]`, and why) are on the threads and in the PR body's review-gate block.

An adversarial pass over the fix filed two follow-ups for gaps it does **not** close: #13544 (a completion is not correlated with the refresh the caller triggered — already open, rediscovered independently) and #13603 (a completion is not revalidated against its table, so a table removed *after* a completion can still be acked ready). Both are named in `RefreshCompletionOutcome`'s doc so `Answered` is not read as more than it claims.

`Attestation` is red until the remote sign-off dispatched for this head returns — the local gate cannot run here, since `crates/cayenne` does not build on Windows and `crates/runtime` depends on it. Nothing else is outstanding: no conflict with `trunk`, no unresolved threads, and the only remaining gate is a maintainer review.

**Review by lukekim — APPROVED** [review]

**Comment by claudespice** [comment]
**Position 1, `UNMERGEABLE` — both failures in this batch are the runner host, not this PR. No change requested.**

The 08:29Z batch (`gh-readonly-queue/trunk/pr-13505-b967366e`) failed two checks:

- **`E2E Test CI`** — `chat_01.exp against a healthy REPL` timed out waiting for the initial chat prompt on `darwin-aarch64`, while `chat_01_simple.exp` against the same REPL passed ten seconds later. Re-run: **green**.
- **`Build and Test`** — 1 hard failure out of 11977 (`5 flaky, 2 leaky`): `runtime_launcher::tests::child_process::a_termination_signal_reaches_the_runtime`. Its inner subprocess panicked with `the stub runtime never created …/started` — a `#!/bin/sh` stub failing to touch one file inside a 10-second budget. That is the readiness budget losing to the host, not a launcher defect.

Attribution is straightforward: this PR changes 21 files under `crates/runtime` and `crates/runtime-table` and touches **nothing** under `bin/spice/`, where the failing test lives, and the identical head passed the whole `pr` workflow in the 06:42Z batch.

The mechanism is measurable on the host: **load average 65.7 on a 10-core machine** (6.5× oversubscribed) with seven remote sign-offs running alongside the merge-queue batches on the same runners. A 10-second readiness budget calibrated on an idle box is a guillotine at that load. One of those sign-offs has been cancelled to give the batch room; the rest are other branches' and were left alone.

Next step is a re-run of `Build and Test` once the batch's `pr` run finishes — it is still in progress on `Rust Lint`. Nothing to fix here.
